src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java

```diff
@@ -449,6 +452,17 @@ private void processIter(DetailAST root, AstState astState) {
         }
     }
 
+    /**
+     * Creates a new {@link SortedSet} with a deterministic order based on the
+     * Check's name before the default ordering.
+     *
+     * @return The new {@link SortedSet}.
+     */
+    private static SortedSet<AbstractCheck> createNewCheckSortedSet() {
+        return new TreeSet<>(Comparator.comparing(AbstractCheck::getClassName)
+                .thenComparing(AbstractCheck::hashCode));
```

Comment:
can we do one more sorting by Id
https://github.com/checkstyle/checkstyle/blob/4d839eb396151fff3f200af06ba99b27ca8eb632/src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractViolationReporter.java#L78-L80

?

Reply:
I need clarification.
I assume this is addition to class name and hash since ID is not always set? Sorting by ID means execution of check will be by ID which is whatever user defines, and will not be controlled by check name or us.
You want to sort by ID before class name?

Reply:
hashcode is not also controlled by us
```
Comparator.comparing(AbstractCheck::getClassName)
                .thenComparing(AbstractCheck::getId))
                .thenComparing(AbstractCheck::hashCode));
```

I hope it make sense.

Reply:
> hashcode ... I am not even think it is useful

It is **required** otherwise duplicate instances with no/same ID will be dropped as if there were never in the config. See https://github.com/checkstyle/checkstyle/pull/9254#discussion_r571646551

Reply:
@romani 
> .thenComparing(AbstractCheck::getId)

This does not work since `id` can be `null`. This isn't `null` safe and creates a bunch of NPEs.

Reply:
Do we have ability to have comparator that is null safe ?

Reply:
@strkkk is this the best way to do it?
````
        return new TreeSet<>(Comparator
                .<AbstractCheck, String> comparing(check -> check.getClass().getName())
                .thenComparing(Comparator.nullsLast(Comparator.comparing(AbstractCheck::getId)))
                .thenComparing(AbstractCheck::hashCode));
````

Reply:
looks like it is. There is also `org.apache.commons.collections.comparators.NullComparator` but it basically does same thing.

I would also consider setting some default value for id, e.g. empty string. It will help to avoid NPEs

Reply:
@rnveach , please do your suggestion.

Reply:
@romani @strkkk I am still getting NPEs. I am lost on how to resolve this.

Reply:
@rnveach , ok , please create separate issue on this 
and lets do without Id sorting.

Reply:
@rnveach you are getting NPEs because last comparator (hashCode) is not null-safe.
So following 
```
return new TreeSet<>(Comparator
                .<AbstractCheck, String> comparing(check -> check.getClass().getName())
                .thenComparing(Comparator.nullsLast(Comparator.comparing(AbstractCheck::getId)))
                .thenComparing(Comparator.nullsLast(Comparator.comparing(AbstractCheck::hashCode)));
```

Reply:
@strkkk `hashCode` returns `int` not an object.

I found https://stackoverflow.com/a/53315210/1016482 and the issue is `nullsLast` is in the wrong location. We were ensuring the object was not `null`, which can never happen here, instead of ensure the result of the method is not `null`.

Reply:
Done.